### ISSUE TITLE
New Collection API RFC

### ISSUE THREAD

**Issue by pzuraq** [san650/ember-cli-page-object#352]
After revisiting #347 I realized that there's a lot going on in that PR and I wanted to clarify the design and intent behind it. The proposal is ultimately a major departure from the existing `collection` API and I think it would be good to discuss it in a semi-RFC like manner.

## Motivation

The current collection API is somewhat difficult to work with for two main reasons:

1. It breaks the mental model of page objects by forcing item definitions into a "scope" which doesn't exist.

    For most pages, users can follow assume that the tree structure of the page definition exactly matches the tree structure of the final page. With collections, this is not true:

    ```js 
    let page = create({
      items: collection({
        scope: 'table', 
        caption: 'caption', 
        itemScope: 'tr', 
        resetScope: true,

        item: {
          foo: 'bar'
        }
      })
    });

    page.items(0).foo; // 'bar' 
    page.items(0).caption; // undefined
    page.items().forEach((i) => i.foo); // 'bar' 
    page.items().forEach((i) => i.caption); // 'caption' 
    page.items().caption // 'caption'
    ```

    As you can see, there is no `item` scope for users to chain off of. Users must also know how to follow the various rules of where values that defined in collections end up:

    * The `scope` that is passed into the collection is not actually related to the items themselves, it is the _container_ of the items. 
    * You must define the item's scope on `itemScope`, not in the item definition itself 
    * You can access other values on the collection definition on the enumerable, and nowhere else (as in the caption example) 
    * `resetScope` must be set at the root collection level, and not the item level. Confusingly, it also applies to _both_ the items _and_ the item container. There is no way to specify `resetScope` for just one or the other.

    This is not a very complex set of rules, but it can be a bit counter-intuitive to newcomers to have to learn these details.

2. It operates as both a query function _and_ an enumerable depending on how users decide to use it, which puts us into different competing mental models. Users can either call the collection as a function to "query" it with a value, or call it without a value to receive back an enumerable and do more complex operations:

    ```js
    page.items(0); page.items().forEach(() => /**/); 
    ```

    This is confusing because we have no way of using the API consistently. Unlike jquery, queries return a single page without any of the collection functions (`$('.foo').each()` and `$('.foo:eq(0)').each()` are both valid, jquery always returns an iterable). The only way to access the enumerable is the special cases where no values are passed, and in turn this is the only way to access special properties like the `caption` from the previous example.

These API inconsistencies ultimately make it difficult to learn how collections work, and difficult to expand on them. 

## Detailed design

This proposal seeks to fix the above issues by promoting the enumerable model of collections to be the only model:

```js 
// before

let page = create({
  items: collection({
    scope: 'table', 
    caption: 'caption', 
    itemScope: 'tr', 
    resetScope: true, 
    
    item: {
      foo: 'bar'
    }
  })
});

page.items(0).foo; // 'bar' 
page.items(0).caption; // undefined
page.items().forEach((i) => i.foo); // 'bar' 
page.items().forEach((i) => i.caption); // 'caption' 
page.items().caption // 'caption' 

// After

let page = create({
  table: {
    scope: 'table', 
    caption: 'caption', 
    resetScope: true,

    items: collection({
      scope: 'tr', 
      resetScope: true, // resetScope previously applied to both parent and child so we must put it in both places to maintain the same semantics 
      foo: 'bar'
    })
  }
});

page.table.caption; // access the caption 
page.table.items.eq(0); // access the first item 
page.table.items.forEach((i) => /**/); // iterate over items 
```

This API solves the above issues by:

1. Flattening definitions. `table` is now its own clearly defined page which is the parent of the collection, and items is clearly a child. The definition for individual items is passed directly into `collection` and it works exactly the same as if it where _not_ being used for a collection. The only difference is that we expect multiple items with the same definition to be in this scope. 

2. Creating a single way of accessing items, through the enumerable. The enumerable is essentially an array wrapper, and users should think of it as an array. We can expand the power of this array to give it more advanced functionality, like `findOne` and `findAll` methods that could quickly filter the collections and provide assertions and warnings. 

As #347 shows, we can begin to adopt this new model incrementally without breaking changes. Unfortunately, flattening definitions will definitely be a breaking change if we want to replace the `collection` helper directly. We could also considered creating a new helper that users could use to transition incrementally. 

## How We Teach This

Aside from updating the collection documentation, we should also be clear that the mental model behind collections is as an array class, much like `Ember.A`. If possible, deprecation warnings for using the old style should be added before breaking changes are made.

## Drawbacks

* This is a breaking change, and would require people using `ember-cli-page-object` to update accordingly. The update process may be codemod-able, so it may not be super difficult to accomplish, but it is still not ideal for users. 
* Adding parts of the new model incrementally as in #347 may be confusing for users, since it will add many new ways of doing things all at once. We would have to be clear that the intent is to transition away from the query-style of collections.

## Alternatives

* Instead of promoting the enumerable model, we could expand the power of querying as suggested in #288. This would make collections more powerful as is, but wouldn't solve many of the issues this RFC brings up. * We could adopt a jquery-like model where queries always return an enumerable.

    ```js 
    page.items(0).forEach(); 
    page.items('.selector').forEach();
    page.items().forEach(); 
    ```

    This feels somewhat awkward and would also be breaking, but does bring the API into alignment across all the ways that collections are query-able. 

## Unresolved questions

* Would we want to make these changes for v2? Or wait until v3, as there will be many other changes happening in the same time frame (native-helpers and testing framework changes coming soon)
* What array methods should be exposed on the enumerable? Should we provide special "query" methods like `findOne` and `findAll`? Does `eq` work for getting the nth item, or should we go with `Ember.A`'s `objectAt` instead?

**Comment by ro0gr** [comment]
I like `item`less collections proposal. I think it can really improve learn-ability.

However it'd prefer to keep backward compatibility. It shouldn't cost to much to leave an `itemScope` support in the `collections` API.

**Comment by ro0gr** [comment]
This part of proposal is much less obvious to me:
```js
table.items.eq(0);
table.items.forEach((i) => /**/); 
```

In my opinion the current way to access an item(`.items(0)`) is much more closer to what I'd really like to write:
```js
table.items[0];
table.items.forEach((i) => /**/); 
```

**Comment by pzuraq** [comment]
I can definitely get behind making `itemScope` an alias for `scope` to ease transitioning, I don’t think that property would be used for anything else commonly. `item`less collections would still be breaking in other ways, unfortunately I don’t think there’s a way for us to truly maintain full backwards compat, but we can definitely make it easier (and I’d be up for writing a codemod!)

The main reason for using an accessor in the `eq` method is because we need to generate pages on the fly, because item counts can change over time and we won’t know which pages we need ahead of time. We could try to keep the current syntax though, and just deprecate calling the function without an index:

```js
table.items(0);
table.items.forEach();
table.items().forEach(); // deprecated, would be removed in the future
```

Basically it works like an array, but with parens instead of brackets

**Comment by pzuraq** [comment]
We could also consider using [proxies](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) for the array wrapper as well to make it work just like a native array, but we wouldn’t be able to support IE11 that way 😕

**Comment by magistrula** [comment]
1. I like your point about bringing `collection` in line with the mental model we're operating under everywhere else in the page object. One result of having `scope` live inside the item definition is that this will be possible: 
```javascript
myItems: collection(SomeImportedPageObjectDefinition)
```
Instead of this, which is what you currently have to do:
```javascript
myItems: collection({
  itemScope: SomeImportedPageObjectDefinition.scope,
  item: SomeImportedPageObjectDefinition
})
```

2. I think it makes sense for changes to `collection` to go into `v3`. That way, continued discussion/development of this idea won't block the release of `v2`. Additionally, it will allow people to address breaking changes more incrementally if needed.

3. I like `objectAt` rather than `eq` for consistency with other Embery-sounding collection methods (`filterBy`, `mapBy`). I agree with @ro0gr that `items(0)` is closer to the `items[0]` that I want to write, but I think it would be a bit weird for `items` to be something that you sometimes call (`items(0)`) and sometimes don't call `items.forEach()`.

**Comment by pzuraq** [comment]
1. I think that example is great, exactly summarizes some of the consistency `item`less collections would bring to writing page objects!

2. I agree that pushing back these changes to v3 is probably for the best. I know we wanted to push for native support by default in v2, and in general catch up to the changes in ember-cli-test-helpers and such. I'm divided on whether or not we should introduce them as a new helper, `collection` is a great name and I'm afraid it could be confusing to new users to have two ways to do the same thing in the meantime. 

    Another strategy I've been thinking about is providing a flag that allows users to opt-in, perhaps through a new addon that could be added by default for new users. The addon just turn on the flag, similar to the way Ember has approached new features and legacy support (like the legacy-views addon).

3. The more I think about it, the more I agree. `objectAt` implies this is an array proxy like Ember.A, whereas `eq` would make me think it was more jquery-like. We could probably implement it using Ember.A if we wanted actually. 

    I also think having two different ways of interacting with it would be strange and probably throw people off at first. It's somewhat more verbose, but for many use cases it's better.

    We actually have

**Comment by ro0gr** [comment]
> I think it makes sense for changes to collection to go into v3

Actually I would like to see `item`less collections API rather sooner then later, if we find out a way to keep backward compat of course.

I think we are far enough from v2 release yet. We don't have anything breaking at the plate. 
The list [to be deprecated](https://github.com/san650/ember-cli-page-object/issues/344) is also not complete. So I think we could afford ourselves to break the world a little bit 🎳 

> I know we wanted to push for native support by default in v2

afaik we don't have an agreement or an actual plan for that.

> collection is a great name.

I agree. It would be hard to compete with this name.

> I can definitely get behind making itemScope an alias for scope to ease transitioning, I don’t think that property would be used for anything else commonly. itemless collections would still be breaking in other ways

I think there is one more option to consider 🤔 
What if we force collection to behave in 2 different ways depending on its options? 
 - If `collection` recognizes `item` or `itemScope` in the options it would do exactly the same as it does today but with a new deprecation. 
 - Otherwise It would choose the new strategy described in this issue 

I'm not very familar with `collection` internals but from my naive point of view it should be totally doable.
I understand that it may be considered as an ugly architectural decision. However it could allow us to introduce this feature even in the 1.x series. 

What do you think guys about this way?

**Comment by pzuraq** [comment]
I think we could mark `item` and `itemScope` special fields that opt you back into the old API, as long as they are required together. I think it's very unlikely that people would want to use both keys together and not want the old API. It does seems like a bit of an ugly architectural decision, but it would also be the only option so far where users could transition one collection at a time, allowing large codebases to transition incrementally.

**Comment by magistrula** [comment]
I like the idea of supporting both the new and old `collection` behavior simultaneously. Agree that it is kinda ugly, but it's such a big win in terms of developer convenience. Supporting both API's will decouple introducing the new API from breaking the old API, which will give us more freedom to choose when it's appropriate to do the latter.
